### PR TITLE
Workstreams Final Fixes

### PR DESCRIPTION
**Pull request by suppermancool**
Workstreams Final Fixes

### CODE REVIEW

File: src/projects/actions/projectPlan.js [hunk 1]
```diff
@@ -54,7 +55,10 @@ function getData(project, dispatch) {
 
   return dispatch(loadProjectPhasesWithProducts(project.id))
     .then(({ value: phases }) => {
-      loadTopics(project.id, _.map(phases, (phase) => `phase#${phase.id}`), dispatch)
+      // load timelines for phase products here together with all dashboard data
+      // as we need to know timeline data not only inside timeline container
+      loadTimelinesForPhasesProducts(phases, dispatch)
+      return loadTopics(project.id, _.map(phases, (phase) => `phase#${phase.id}`), dispatch)
```
Comment (maxceem) [inline src/projects/actions/projectPlan.js]: @suppermancool actually when we are showing phases it's ok to show them without waiting for loading of comments to show the phases faster. Is this change is required for some other functionality?

File: src/projects/actions/works.js [hunk 2]
```diff
@@ -303,7 +307,9 @@ export function loadWorkitems(projectId, workstreamId, workId) {
       type: LOAD_WORK_ITEM,
       payload: getWorkitems(projectId, workstreamId, workId)
         .then((results) => {
-          Promise.all([loadChallengesForWorkItems(results, dispatch)])
+          if (results.length > 0) {
+            Promise.all([loadChallengesForWorkItems(results, dispatch)])
```
Comment (maxceem) [inline src/projects/actions/works.js]: Can we directly call `loadChallengesForWorkItems(results, dispatch)` without `Promise.all` here?

File: src/projects/detail/components/workstreams/WorkListCard.jsx [hunk 3]
```diff
@@ -33,7 +31,11 @@ const WorkListCard = (props) => {
       return (
         <button
           className="tc-btn tc-btn-primary tc-btn-sm"
-          onClick={() => inputDesignWorks(timeline.id, activeMilestone.id)}
+          onClick={(e) => {
+            e.preventDefault()
+            // Start input design works for timeline
+            props.history.push(`${workDashboardUrl}/milestones/${activeMilestone.id}`)
+          }}
```
Comment (maxceem) [inline src/projects/detail/components/workstreams/WorkListCard.jsx]: Can we use `<Link>` here instead of button with custom `onClick` logic as it's done in `ActiveMilestoneSummary.jsx `.

File: src/projects/detail/components/workstreams/WorkListCard.jsx [hunk 4]
```diff
@@ -46,7 +48,7 @@ const WorkListCard = (props) => {
           onClick={(e) => {
             e.preventDefault()
             // Start design review for timeline
-            props.history.push(`${workDashboardUrl}/timelines/${timeline.id}/milestones/${activeMilestone.id}/review`)
+            props.history.push(`${workDashboardUrl}/milestones/${activeMilestone.id}`)
```
Comment (maxceem) [inline src/projects/detail/components/workstreams/WorkListCard.jsx]: To keep it consistent, can we use `<Link>` also here instead of button with custom `onClick` logic as it's done in `ActiveMilestoneSummary.jsx `.

File: src/projects/detail/components/workstreams/WorkViewRequirements.jsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
+/**
+ * WorkListCard section
```
Comment (maxceem) [inline src/projects/detail/components/workstreams/WorkViewRequirements.jsx]: Comment doesn't match component name.

File: src/projects/detail/components/workstreams/WorkViewRequirements.jsx [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+  componentWillReceiveProps(nextProps) {
+    const {
+      progressId,
+    } = nextProps
+
+    // close edit feedback form
+    const prevIsUpdatingWorkInfoWithProgressId = _.get(this.props, `isUpdatingWorkInfoWithProgressId.${progressId}`)
+    const nextIsUpdatingWorkInfoWithProgressId = _.get(nextProps, `isUpdatingWorkInfoWithProgressId.${progressId}`)
+    if (prevIsUpdatingWorkInfoWithProgressId === true && nextIsUpdatingWorkInfoWithProgressId === false) {
```
Comment (maxceem) [inline src/projects/detail/components/workstreams/WorkViewRequirements.jsx]: We also should check that no error happened during sending request to the server. Otherwise user will loose all the changes and editor would be closed. To reproduce:
- open requirements editor
- edit requirements
- disable internet
- click "Done"
- the requests to the server will fail, but editor would be closed and we will loose all the changes.
- so we should keep editor opened if there is error during saving to the server

File: src/projects/detail/containers/WorkAssetsContainer.jsx [hunk 7]
```diff
@@ -84,11 +87,28 @@ class WorkAssetsContainer extends React.Component {
...
+  for (const key in topics) {
+    // check also if property is not inherited from prototype
+    if (topics.hasOwnProperty(key)) {
+      const topic = topics[key]
+      if (topic.isLoading) {
+        isLoadingTopic = true
+        break
+      }
+    }
+  }
```
Comment (maxceem) [inline src/projects/detail/containers/WorkAssetsContainer.jsx]: It looks like this code can be replaced using `lodash` by 
```js
const isLoadingTopic = _.some(_.values(topics), { isLoading: true })
``` 
I would prefer not to use `in` with `hasOwnProperty` as it's too wordy and easy to forget about `hasOwnProperty`.

File: src/projects/detail/containers/WorkTimelineMilestoneContainer.jsx [hunk 8]
```diff
@@ -0,0 +1,122 @@
+/**
+ * WorkTimelineMilestoneContainer container
+ * displays content of the work milestone review section
```
Comment (maxceem) [inline src/projects/detail/containers/WorkTimelineMilestoneContainer.jsx]: > displays content of the work milestone review section

It looks like it displays content of any work milestone, not only `review section`

File: src/projects/detail/containers/WorkTimelineMilestoneContainer.jsx [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+          <LoadingIndicator />
+        </div>
+      )
+    }
+
+    switch (milestone.type) {
+    case MILESTONE_TYPE.DESIGN_WORK:
+      return (
+        <DesignWorksContainer
+          onBack={() => this.setState({ showDesignWorks : {timelineId: -1, milestoneId: -1} })}
```
Comment (maxceem) [inline src/projects/detail/containers/WorkTimelineMilestoneContainer.jsx]: I guess this `onBack` is not in use anymore.

File: src/projects/actions/projectPlan.js (hunk 1 above)
Comment (suppermancool) [reply src/projects/actions/projectPlan.js]: Actually there is requirement for assets page, i will update PR to make this only for Assets page
> `Also, we should show the loading indicator on Assets Library page until all the posts are loaded.`

Comment (suppermancool) [reply src/projects/actions/projectPlan.js]: @maxceem done in the latest commit

File: src/projects/actions/works.js (hunk 2 above)
Comment (suppermancool) [reply src/projects/actions/works.js]: @maxceem done in the latest commit

File: src/projects/detail/components/workstreams/WorkListCard.jsx (hunk 4 above)
Comment (suppermancool) [reply src/projects/detail/components/workstreams/WorkListCard.jsx]: @maxceem done in the latest commit

File: src/projects/detail/components/workstreams/WorkViewRequirements.jsx (hunk 5 above)
Comment (suppermancool) [reply src/projects/detail/components/workstreams/WorkViewRequirements.jsx]: @maxceem done in the latest commit

File: src/projects/detail/components/workstreams/WorkViewRequirements.jsx (hunk 6 above)
Comment (suppermancool) [reply src/projects/detail/components/workstreams/WorkViewRequirements.jsx]: @maxceem done in the latest commit

File: src/projects/detail/containers/WorkTimelineMilestoneContainer.jsx (hunk 8 above)
Comment (suppermancool) [reply src/projects/detail/containers/WorkTimelineMilestoneContainer.jsx]: @maxceem done in the latest commit

File: src/projects/detail/containers/WorkTimelineMilestoneContainer.jsx (hunk 9 above)
Comment (suppermancool) [reply src/projects/detail/containers/WorkTimelineMilestoneContainer.jsx]: @maxceem done in the latest commit

File: src/projects/detail/containers/WorkAssetsContainer.jsx (hunk 7 above)
Comment (suppermancool) [reply src/projects/detail/containers/WorkAssetsContainer.jsx]: @maxceem done in the latest commit

File: src/projects/detail/components/workstreams/WorkListCard.jsx (hunk 3 above)
Comment (suppermancool) [reply src/projects/detail/components/workstreams/WorkListCard.jsx]: @maxceem done in the latest commit

File: src/projects/detail/components/work-timeline/WorkTimelineEditMilestone.jsx [hunk 10]
```diff
@@ -264,7 +264,7 @@ class WorkTimelineEditMilestone extends React.Component {
                 label="Duration (days)"
                 type="number"
                 name="duration"
-                value={duration}
+                value={`${duration}`}
```
[the remark this answers is not in the document]
Comment (suppermancool) [reply src/projects/detail/components/work-timeline/WorkTimelineEditMilestone.jsx]: @maxceem 
There is validator rule for this TextInput:
`
value={`${duration}`}
minValue={1}
// cannot use this validator because of the issue in `isRequired` validator for numbers
// see issue
validations={{
                  isRequired: true
                }}
`
isRequired only work for string(it throw error for me because duration is number) so i convert duration to string. Actually i can't open the edit milestone form because this error, you can try to revert the code and check if it work for you(maybe this only happen on my pc). FYI: after revert duration to string it still work like before.

[the remark this answers is not in the document]
Comment (maxceem) [reply src/projects/detail/components/work-timeline/WorkTimelineEditMilestone.jsx]: Thanks, got it. This has been fixed in `react-components` library. So to get working without making it a string, have to remove `node_modules` and reinstall dependencies.

Thanks for letting me know, so I'll revert this change.
